### PR TITLE
Data tag

### PR DESCRIPTION
**Pull request by willghatch**
Is this the direction you're looking for, Jay?

Right now the DLE_DataTag has a major flaw in that it doesn't check which `Data` type something is, so this version of `isSome` doesn't either.  I thought “I'll fix this with the `is` function”, but `Maybe` is a type constructor and not a type, so I can't use `is(x, Maybe)` and there is no concrete type that I can pass in there to be correct for `isSome`.

### CODE REVIEW

File: hs/rsh/stdlib.rsh [hunk 1]
```diff
@@ -35,8 +35,10 @@ export const fromMaybe = (v, onNull, onSome) => {
   case None: return onNull();
   case Some: return onSome(v); } };
 
-export const isSome = m => fromMaybe(m, () => false, (_) => true);
-export const isNone = m => fromMaybe(m, () => true, (_) => false);
+// TODO - this definitely emits optimized teal code, but it no longer checks that the data is actually a proper Maybe type...
```
Comment (chrisnevers) [inline hs/rsh/stdlib.rsh:38]: Maybe a silly idea, but what about this for typechecking?
```
const isSome = m => {
  assert(m.match({ Some: true, None: true });
 return _dataTag(m) == 1;
}

File: hs/rsh/stdlib.rsh [hunk 2]
```diff
@@ -35,8 +35,10 @@ export const fromMaybe = (v, onNull, onSome) => {
...
+// TODO - also optimize isLeft/isRight, etc
+export const isSome = m => _dataTag(m) === 1;
+export const isNone = m => _dataTag(m) === 0;
```
Comment (jeapostrophe) [inline hs/rsh/stdlib.rsh:41]: Idea: The new thing is `DLE_DataTag` but the SL feature is `isDataVariant(variantName, otherVariants, instance)` and the type checking happens just in SL but compiles to DL

File: hs/src/Reach/Connector/ALGO.hs [hunk 3]
```diff
@@ -2343,6 +2343,10 @@ ce = \case
         -- [ extra ]
         return ()
     label after_lab
+  DLE_DataTag _ d -> do
+    ca d
+    cint 0
+    op "getbyte"
```
Comment (jeapostrophe) [inline hs/src/Reach/Connector/ALGO.hs:2349]: Double check if this is really a number

File: hs/src/Reach/Connector/ETH_Solidity.hs [hunk 4]
```diff
@@ -730,6 +731,10 @@ solExpr sp = \case
   DLE_TimeOrder {} -> impossible "timeorder"
   DLE_EmitLog {} -> impossible "emitLog"
   DLE_setApiDetails {} -> impossible "setApiDetails"
+  DLE_DataTag _ d -> do
+    d' <- solArg d
+    -- TODO - what kind of number should this be converted into?
```
Comment (jeapostrophe) [inline hs/src/Reach/Connector/ETH_Solidity.hs:736]: `uint256`

File: hs/src/Reach/Eval/Core.hs [hunk 5]
```diff
@@ -624,6 +624,7 @@ base_env_slvals =
   , ("call", SLV_Form $ SLForm_apiCall)
   , (".setApiDetails", SLV_Form $ SLForm_setApiDetails)
   , ("getUntrackedFunds", SLV_Prim $ SLPrim_getUntrackedFunds)
+  , ("_dataTag", SLV_Prim $ SLPrim_DataTag)
```
Comment (jeapostrophe) [inline hs/src/Reach/Eval/Core.hs:627]: If it really were this feature, it would be `unsafeDataTag`

File: hs/src/Reach/Eval/Core.hs [hunk 6]
```diff
@@ -3798,6 +3799,19 @@ evalPrim p sargs =
       dv <- ctxt_lift_expr mdv $ DLE_PrimOp at (ADD UI_Word) [DLA_Var untrackedFunds, trackedBal]
       setBalance TM_Balance mtok (DLA_Var dv)
       return (lvl, SLV_DLVar untrackedFunds)
+    SLPrim_DataTag -> do
+      v <- one_arg
+      at <- withAt id
+      case v of
+        SLV_Data _ dmap vn _ -> do
+          let tm = M.fromAscList $ zip (map fst $ M.toAscList dmap) ([0 ..] :: [Integer])
```
Comment (jeapostrophe) [inline hs/src/Reach/Eval/Core.hs:3807]: I think this is the function you defined

File: hs/src/Reach/Verify/SMT.hs [hunk 7]
```diff
@@ -1190,6 +1190,17 @@ smt_e at_dv mdv de = do
...
+      d' <- smt_a at d
+      n <- smtTypeSort $ argTypeOf d
+      let variants = map fst $ M.toAscList $ dataTypeMap $ argTypeOf d
+      let variantCtors = map (\v -> n <> "_" <> v) variants
+      let variantVs = map (\vCtor -> Atom $ vCtor <> "_pv") variantCtors
+      let variantPs = map (\(vV, vC) -> List [Atom vC, vV])
+           $ zip variantVs variantCtors
+      let variantCs = map (\(vP, i) -> List [vP, Atom $ show i])
+           $ zip variantPs ([0 ..] :: [Int])
+      bound at $ smtApply "match" [d', List variantCs]
```
Comment (jeapostrophe) [inline hs/src/Reach/Verify/SMT.hs:1203]: This should probably be a function that gets defined with the data type and then we just call it

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

File: hs/src/Reach/Connector/ALGO.hs (hunk 3 above)
Comment (willghatch) [reply hs/src/Reach/Connector/ALGO.hs:2349]: I'm not sure I understand this feedback.  Do you mean “double check that getbyte returns a number”?  If so, the docs say that it returns an integer.

Comment (jeapostrophe) [reply hs/src/Reach/Connector/ALGO.hs:2349]: Yea. That's what I meant

File: hs/rsh/stdlib.rsh (hunk 1 above)
Comment (willghatch) [reply hs/rsh/stdlib.rsh:38]: Thanks!  Good idea, although superseded by Jay's suggestion.

**Review by jeapostrophe — CHANGES_REQUESTED** [review]
Docs, Changelog, tests (static & dynamic)

**Comment by jeapostrophe** [comment]
Excellent

**Comment by willghatch** [comment]
The failing test looks like it gets a slightly different witness on my machine, thus the golden file doesn't match (because I changed it to what I see, and it looks like the CI is getting a different message).  I bet this is just my version of Z3 being out of sync (I see it was just updated).

File: docs/src/changelog/index.md [hunk 8]
```diff
@@ -6,6 +6,7 @@ Versions and changes-within-versions are listed in reverse-chronological order:
 ## 0.1.11: 2022/06 - present
 
 @{verRC("0.1.11")}
++ 2022/08/03: Added support for `{!rsh} isDataVariant`.
```
Comment (jeapostrophe) [inline docs/src/changelog/index.md:9]: This should be under rc6 (but actually, you will be beat and it will end up being in rc7)

Comment (jeapostrophe) [reply docs/src/changelog/index.md:9]: Also, just say "Added ..."

File: examples/isDataVariant/index.rsh [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    getDT: Fun([], DT),
+  });
+  init();
+
+  A.only(() => {
+    const x = declassify(interact.getDT());
+  });
+  A.publish(x);
+  const isDV = isDataVariant("A", ["A", "B"], x);
+  enforce(isDV);
```
Comment (jeapostrophe) [inline examples/isDataVariant/index.rsh:16]: This is an immoral use of `enforce` :)

File: docs/src/rsh/compute/index.md [hunk 10]
```diff
@@ -1361,6 +1361,13 @@ As a special case, when the type of a variant is `{!rsh} Null`, the `{!rsh} VALU
 
 Data instances are consumed by `{!rsh} switch` statements and `{!rsh} match` expressions.
 
+@{ref("rsh", "isDataVariant")}
+Data instances may be checked against particular variants with the `{!rsh} isDataVariant` function.
+
+```reach
+isDataVariant("Circle", ["Circle", "Square", "Rect"], nice); // true
+```
+
```
Comment (jeapostrophe) [inline docs/src/rsh/compute/index.md:1370]: You could say "This is like the following switch... but more efficient"

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

File: examples/isDataVariant/index.rsh (hunk 9 above)
Comment (willghatch) [reply examples/isDataVariant/index.rsh:16]: Bleh, it's the easiest way to be sure this kind of test code isn't optimized out of a consensus step :p

Comment (jeapostrophe) [reply examples/isDataVariant/index.rsh:16]: :hankey:

**Review by jeapostrophe — APPROVED** [review]

**Pull request merged**